src/modals/WalletConnect/WalletConnectSessionProposalModal.tsx

```diff
@@ -0,0 +1,200 @@
+import { selectAddressesInGroup } from '@/storage/addresses/addressesSelectors'
+import { saveNewAddresses } from '@/storage/addresses/addressesStorageUtils'
+import { networkPresetSwitched } from '@/storage/settings/networkActions'
+import { Address } from '@/types/addresses'
+import { SessionProposalEvent } from '@/types/walletConnect'
+import { getRandomLabelColor } from '@/utils/colors'
+import { cleanUrl } from '@/utils/misc'
+import { isNetworkValid, parseSessionProposalEvent } from '@/utils/walletConnect'
+
+interface WalletConnectSessionProposalModalProps {
+  approveProposal: (signerAddress: Address) => Promise<void>
+  rejectProposal: () => Promise<void>
+  proposalEvent: SessionProposalEvent
+  onClose: () => void
+}
+
+const WalletConnectSessionProposalModal = ({
+  onClose,
+  approveProposal,
+  rejectProposal,
+  proposalEvent
+}: WalletConnectSessionProposalModalProps) => {
+  const { t } = useTranslation()
+  const posthog = usePostHog()
+  const currentNetworkId = useAppSelector((s) => s.network.settings.networkId)
+  const currentNetworkName = useAppSelector((s) => s.network.name)
+  const dispatch = useAppDispatch()
+  const { requiredChainInfo, metadata } = parseSessionProposalEvent(proposalEvent)
+  const addressesInGroup = useAppSelector((s) => selectAddressesInGroup(s, requiredChainInfo?.addressGroup))
+  const { generateAddress } = useAddressGeneration()
+
+  const [signerAddress, setSignerAddress] = useState<Address>()
+
+  const group = requiredChainInfo?.addressGroup
+
+  const showNetworkWarning =
+    requiredChainInfo?.networkId && !isNetworkValid(requiredChainInfo.networkId, currentNetworkId)
+
+  useEffect(() => {
+    setSignerAddress(
+      addressesInGroup.length > 0 ? addressesInGroup.find((a) => a.isDefault) ?? addressesInGroup[0] : undefined
+    )
+  }, [addressesInGroup])
+
+  const handleSwitchNetworkPress = async () => {
+    if (requiredChainInfo?.networkId === 'mainnet' || requiredChainInfo?.networkId === 'testnet') {
+      dispatch(networkPresetSwitched(requiredChainInfo?.networkId))
+    }
+  }
+
+  const generateAddressInGroup = () => {
+    const address = generateAddress({ group })
+    saveNewAddresses([{ ...address, isDefault: false, color: getRandomLabelColor() }])
+
+    posthog.capture('New address created through WalletConnect modal')
+  }
+
+  const rejectAndCloseModal = async () => {
+    rejectProposal()
+    onClose()
+
+    posthog.capture('Rejected WalletConnect connection by closing modal')
+  }
+
+  return (
+    <CenteredModal
+      title={
+        metadata.description ? t('Connect to {{ dAppUrl }}', { dAppUrl: cleanUrl(metadata.url) }) : t('Connect to dApp')
+      }
+      subtitle={metadata.description || metadata.url}
+      onClose={rejectAndCloseModal}
+      Icon={() =>
+        metadata?.icons &&
+        metadata.icons.length > 0 &&
+        metadata.icons[0] && <AssetLogo assetImageUrl={metadata.icons[0]} assetId={metadata.url} size={50} />
```

Comment:
Nit: If `metadata.icons.length > 0` is `true` then `metadata.icons[0]` must exist?

Reply:
But there is no guarantee that it's not an empty string...

Reply:
Yes but if it's an empty string, then AssetLogo would show a "?" which would be ok?

It's a nit anyway.